odc/geo/_xr_interop.py

```diff
@@ -245,6 +245,59 @@ def assign_crs(
     return xx
 
 
+def crop(xx: XrT, poly: Geometry, mask: bool = True, all_touched: bool = True) -> XrT:
+    """
+    Crops and optionally mask an xr.Dataset or xr.DataArray (``xx``) to
+    the spatial extent of a geometry (``poly``).
+
+    :param xx:
+       :py:class:`~xarray.Dataset` or :py:class:`~xarray.DataArray`.
+
+    :param poly:
+       Geometry shape used to crop ``xx``.
+
+    :param poly:
+       Whether to mask out pixels outside of the rasterized extent of
+       ``poly``.
+
+    :param all_touched:
+        If ``True``, all pixels touched by ``poly`` will remain unmasked.
+        If ``False``, only pixels whose center is within the polygon or
+        that are selected by Bresenham's line algorithm will remain unmasked.
+
+    :return:
+        A :py:class:`~xarray.Dataset` or :py:class:`~xarray.DataArray`
+        cropped and optionally masked to the spatial extent of ``poly``.
+    """
+    # Reproject `poly` to match `xx` and calculate bounding box
+    poly = poly.to_crs(crs=xx.odc.crs)
+    bbox = poly.boundingbox
+
+    # Verify that `poly` overlaps with `xx` extent
+    if poly.intersects(xx.odc.geobox.extent):
+        # Crop `xx` to the bounding box of `poly`
+        ydim, xdim = spatial_dims(xx)
+        xx_cropped = xx.sel(
+            {xdim: slice(bbox.left, bbox.right), ydim: slice(bbox.top, bbox.bottom)}
+        )
```

Comment:
I think this is best covered by `GeoBox.enclosing` followed by geobox intersect, this should give you an roi of the original image overlapping with the geometry, which you can then use with `.isel` to select a sub-region.

Reply:
Combination of these two should give you a slice into original image covering given geometry:

https://github.com/opendatacube/odc-geo/blob/6ce7524b6808457418b2c8bcecf01c918ce6458a/odc/geo/geobox.py#L620

https://github.com/opendatacube/odc-geo/blob/6ce7524b6808457418b2c8bcecf01c918ce6458a/odc/geo/geobox.py#L657

Reply:
in fact we should overload `overlap_roi` to allow geometry on input.

Reply:
Something like this?
```
xx.odc.geobox.overlap_roi(xx.odc.geobox.enclosing(poly))
>>> (slice(118, 285, None), slice(132, 276, None))
```

The output tuple doesn't include coordinate names... do I still need to grab them with `spatial_dims` or can I get this in to `isel` directly?

Reply:
This works, but would be nice to avoid the `spatial_dims` step:
```
    # Crop `xx` to the bounding box of `poly`
    ydim, xdim = spatial_dims(xx)
    roi = xx.odc.geobox.overlap_roi(xx.odc.geobox.enclosing(poly))
    xx_cropped = xx.isel({ydim: roi[0], xdim: roi[1]})
```

Reply:
Ah... the problem is that you can only use an ROI to index a `xr.DataArray`, not an `xr.Dataset`...

![image](https://github.com/opendatacube/odc-geo/assets/17680388/260ae526-c358-4f5a-9893-baa38ee08312)

Reply:
instead of dimension names there is `.odc.ydim` which tells you which tells you where `y,x` dims are, always next to each other, from that one can construct `roi` object that slices to all dimensions.

Reply:
This is probably common enough functionality that needs to be exposed on `.odc` accessor

* Input: 2d slice into spatial dimensions 
* Output:
   * Nd slice into data variable(s) with extra dimensions filled with `:` slices
   * New Dataset/DataArray sliced into geo dims

So `np.s_[100:200, :30] -> np.s_[:, 100:200, :30]` in the common case, but for rgb data it would be `-> np.s_[:, 100:200, :30, :]`.